File: cpp/JSONRPC.cpp
#include "JSONRPC.h"
#include "RPCTransport.h"
#include "RPCRuntime.h"
#include "ppapi/cpp/var.h"
#include "ppapi/cpp/var_array.h"
#include "ppapi/cpp/var_dictionary.h"
#include "RPCRequest.h"

#include <string>
namespace pprpc{
JSONRPC::JSONRPC() {
}

JSONRPC::JSONRPC(RPCTransport* transport) {
	setTransport(transport);
}

JSONRPC::JSONRPC(RPCTransport* transport, RPCRuntime* runtime){
	setTransport(transport);
	setRuntime(runtime);
}

void JSONRPC::setTransport(RPCTransport* transport) {
	this->transport = transport;
	transport->setJSONRPC(this);
}


void JSONRPC::setRuntime(RPCRuntime* runtime) {
	this->runtime = runtime;
}

JSONRPC::~JSONRPC() {
}

bool JSONRPC::is_basic_json_rpc(const pp::Var& obj) {
	if(obj.is_dictionary()){
		pp::VarDictionary d(obj);
		return ((d.Get("jsonrpc").is_string() && d.Get("jsonrpc").AsString() == "2.0")
				&& ((d.Get("method").is_string()
						|| (d.HasKey("id") && (d.HasKey("result")
								|| d.Get("error").is_dictionary())))));
	}
	return false;
}

bool JSONRPC::ValidateRPCRequest(const pp::Var& requestObj) {
	if(is_basic_json_rpc(requestObj)){
		pp::VarDictionary d(requestObj);
		pp::Var id(d.Get("id"));
		return d.Get("method").is_string() && (id.is_undefined() || id.is_int() || id.is_string() || id.is_null());
	}
	return false;
}

RPCRequest JSONRPC::ExtractRPCRequest(const pp::Var& requestObj) {
	pp::VarDictionary dict(requestObj);
	RPCRequest request(dict);
	if(ValidateRPCRequest(requestObj)){
		request.setValid(true);
	} else {
		request.setValid(false);
	}
	return request;
}

bool JSONRPC::ValidateRPCCallback(const pp::Var& callbackObj) {
	if(is_basic_json_rpc(callbackObj)){
		pp::VarDictionary d(callbackObj);
		return d.HasKey("id") && d.HasKey("result") && !d.HasKey("error");
	}
	return false;
}

bool JSONRPC::ValidateRPCError(const pp::Var& errorObj) {
	if(is_basic_json_rpc(errorObj)){
		pp::VarDictionary d(errorObj);
		if(d.HasKey("id") && d.Get("error").is_dictionary() && !d.HasKey("result")){
			pp::VarDictionary e(d.Get("error"));
			return e.HasKey("code") && e.HasKey("message");
		}
	}
	return false;
}

void JSONRPC::HandleRPC(const pp::Var& rpcObj) {
	RPCRequest r = ExtractRPCRequest(rpcObj);
	if(r.isValid){
		// todo call runtime
	} else if(ValidateRPCCallback(rpcObj)){
		// todo call runtime

	} else if(ValidateRPCError(rpcObj)){
		// todo call runtime
	} else {
		// not a JSON RPC.
	}
}

RPCRequest JSONRPC::ConstructRPCRequest(std::string& method,
		unsigned int id, const pp::VarArray& params) {
	return RPCRequest(method, params, id);
}

RPCRequest JSONRPC::ConstructRPCRequest(std::string& method, const pp::VarArray& params) {
	return RPCRequest(method, params);
}

RPCRequest JSONRPC::ConstructRPCRequest(std::string& method) {
	return RPCRequest(method);
}

pp::VarDictionary JSONRPC::ConstructRPCCallback(unsigned int id,
		pp::Var& result) {
	pp::VarDictionary obj;
	obj.Set("jsonrpc","2.0");
	obj.Set("result",result);
	obj.Set("id",(int)id);
	return obj;
}

pp::VarDictionary JSONRPC::ConstructRPCError(unsigned int id, int code,
		std::string& message, const pp::Var& data) {
	pp::VarDictionary obj = ConstructRPCError(id, code, message);
	pp::VarDictionary error_obj = pp::VarDictionary(obj.Get("error"));
	error_obj.Set("data", data);
	obj.Set("error", error_obj);
	return obj;
}

pp::VarDictionary JSONRPC::ConstructRPCError(unsigned int id, int code,
		std::string& message) {
	pp::VarDictionary obj;
	obj.Set("jsonrpc","2.0");
	pp::VarDictionary obj_error;
	obj_error.Set("code", code);
	obj_error.Set("message", message);
	obj.Set("error",obj_error);
	obj.Set("id",(int)id);

	return obj;
}


bool JSONRPC::SendRPCRequest(const pp::Var& rpcObj) {
	if(ValidateRPCRequest(rpcObj) && transport){
		transport->PostMessage(rpcObj);
		return true;
	} else {
		return false;
	}
}

bool JSONRPC::SendRPCCallback(const pp::Var& rpcObj) {
	if(ValidateRPCCallback(rpcObj) && transport){
		transport->PostMessage(rpcObj);
		return true;
	} else {
		return false;
	}
}

bool JSONRPC::SendRPCError(const pp::Var& rpcObj) {
	if(ValidateRPCError(rpcObj) && transport){
		transport->PostMessage(rpcObj);
		return true;
	} else {
		return false;
	}
}


} /*namespace pprpc*/

File: scripts/JSONRPC.js
define(['RPCTransport', 'lodash', "TagLogger"], function(RPCTransport, _, TagLogger){
  var logger = new TagLogger("JSONRPC");
	function JSONRPC(rpcTransport, runtime){
    if(!_.isUndefined(rpcTransport)){
      this.transport = rpcTransport;
      rpcTransport.setJSONRPC(this);
    }

    if(!_.isUndefined(runtime)){
      this.runtime = runtime;
      runtime.setJSONRPC(this);
    }
	}

  function isBasicJSONRPC(rpcObject){
    return _.isObject(rpcObject) && _.isString(rpcObject.jsonrpc) &&
           (rpcObject.jsonrpc == "2.0") &&
           (_.isString(rpcObject.method) || !_.isUndefined(rpcObject.result) || _.isObject(rpcObject.error));
  }

  JSONRPC.prototype.getModule = function(){
    return this.transport.getModule();
  };

  JSONRPC.prototype.setTransport = function(transport){
    this.transport = transport;
  };

  JSONRPC.prototype.setRuntime = function(runtime){
    this.runtime = runtime;
  };

  /**
   * Checks that the given object is a valid JSON RPC request,
   * as defined in the specification http://www.jsonrpc.org/specification#request_object
   * @param rpcObject
   * @returns Boolean True if valid json rpc request, false otherwise.
   */
  JSONRPC.prototype.validateRPCRequest = function(rpcObject){
    var method = rpcObject.method,
        id     = rpcObject.id;

    return isBasicJSONRPC(rpcObject) &&           // the object needs to be a json-rpc 2.0 object.
           _.isString(method) &&                  // the method has to be defined and a string
          ( _.isUndefined(id) ||                  // the id could be left out, or:
          (_.isNumber(id) && id === (id|0)) ||    // the id could be a number that is an integer
          _.isString(id) ||                       // the id could be a string
          _.isNull(id));                          // the id could be 'null'.
  };

  /**
   * Checks that the given object is a valid JSON RPC callback,
   * as defined in the specification http://www.jsonrpc.org/specification#response_object
   * Note that a callback is different from a response. Responses could be callbacks or errors.
   * @param rpcObject
   * @returns boolean
   */
  JSONRPC.prototype.validateRPCCallback = function(rpcObject){
    return isBasicJSONRPC(rpcObject) &&          // must be a json rpc 2.0 object
           !_.isUndefined(rpcObject.result) &&   // a callback must be a result
           _.isUndefined(rpcObject.error) &&     // a callback can't be an error
           !_.isUndefined(rpcObject.id);         // a callback must have an id
  };

  /**
   * Checks that the given object is a valid JSON RPC error,
   * as defined in the specification http://www.jsonrpc.org/specification#response_object
   * @param rpcObject
   * @returns boolean
   */
  JSONRPC.prototype.validateRPCError = function(rpcObject){
    var errorObject = rpcObject.error;
    return isBasicJSONRPC(rpcObject) &&          // must be a json rpc 2.0 object
           !_.isUndefined(rpcObject.id) &&       // must have an id
           _.isObject(errorObject) &&            // must have an error object
           _.isUndefined(rpcObject.result) &&    // can't have a result
           !_.isUndefined(errorObject.code) &&   // must have an error code
           !_.isUndefined(errorObject.message);  // must have an error message
  };

  /**
   * Handles onMessage events by checking whether it is an rpc call, a callback or an error.
   * Calls RPC Runtime methods.
   * @returns {boolean}
   * @param eventObject
   */
  JSONRPC.prototype.handleRPC = function(eventObject){
    var rpcObject = eventObject.data;

    if(this.validateRPCCallback(rpcObject)){
      // it's a successful response that is a callback
      if(!_.isUndefined(this.runtime)){
        this.runtime.handleCallback(rpcObject);
        return true;
      } else {
        console.error("JSONRPC received callback response but no runtime is set to handle it.");
        return false;
      }

    } else if(this.validateRPCError(rpcObject)){
      // it's a "successful" response that is an error
      if(!_.isUndefined(this.runtime)){
        this.runtime.handleError(rpcObject);
        return true;
      } else {
        console.error("JSONRPC received error response but no runtime is set to handle it.");
        return false;
      }
    } else if (this.validateRPCRequest(rpcObject)){
      // it's a rpc call
      logger.debug("Received request", rpcObject);
      if(!_.isUndefined(this.runtime)){
        this.runtime.handleRequest(rpcObject);
        return true;
      } else {
        console.error("JSONRPC received rpc request but no runtime is set to handle it.");
        return false;
      }
    } else{
      // it's not a rpc response.
      logger.warn("Received non-rpc message", rpcObject);
      return false;
    }
  };

  /**
   * Validates that the object that is about to be sent is a valid json rpc request. If it is, the message is sent.
   * @param rpcObject
   * @returns {*}
   */
  JSONRPC.prototype.sendRPCRequest = function(rpcObject){
    if(!_.isUndefined(this.transport)){
      if(this.validateRPCRequest(rpcObject)){
        return this.transport.send(rpcObject);
      } else {
        logger.warn("Failed to send rpc object because it is invalid");
        // might be better doing an error callback.
        return false;
      }
    } else {
      throw new Error("Transport required for sending json rpc");
    }

  };

  JSONRPC.prototype.constructRPCRequest = function(method, id, params){
    return {
      "jsonrpc": "2.0",
      "method": method,
      "id": id,
      "params": params
    };
  };

  JSONRPC.prototype.constructRPCCallback = function(id, result){
    return {
      "jsonrpc": "2.0",
      "id": id,
      "result": result
    };
  };

  JSONRPC.prototype.constructRPCError = function(id, code, message, data){
    return {
      "jsonrpc": "2.0",
      "id": id,
      "error": {
        "code": code,
        "message": message,
        "data": data
      }
    };
  };

	return JSONRPC;
});
